### PR TITLE
Introducing the Backbone.Router to make resources linkable.

### PR DESCRIPTION
**Pull request by aldipower**
New routes are 'apps', 'deployments', 'apps/:appid(/configuration), 'about', 'newapp'.
Introducing a Collection with fetching status.

Pull request on a clean basis from fresh master. Only with related changes to #325.
Comments on #990 recieved attention.

### CODE REVIEW

File: src/main/resources/assets/js/models/Collection.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
+var Backbone = require("backbone");
+
+module.exports = Backbone.Collection.extend({
```
Comment (mlunoe) [inline src/main/resources/assets/js/models/Collection.js]: I think you can come around this by adding a general function in Marathon.jsx like this:
```
updatePolling: function () {
  var modalClass = this.state.modalClass;
  var id = this.state.activeTabId;
  // logic for checking the state and poll based on that
  },
```
and replace these two with `this.updatePolling()`:
https://github.com/mesosphere/marathon/pull/1066/files#diff-6876cc8f56c80c7276969fd16e24b309R136
https://github.com/mesosphere/marathon/pull/1066/files#diff-6876cc8f56c80c7276969fd16e24b309R146

And move this logic to the `updatePolling` function:
https://github.com/mesosphere/marathon/pull/1066/files#diff-6876cc8f56c80c7276969fd16e24b309L381

That way we can set the state and when the component updates (after setting the state) it will update the polling based on the new state. Does that make sense?

Comment (mlunoe) [inline src/main/resources/assets/js/models/Collection.js]: In general this should ensure that you kill of any ongoing polls before you start the new one:
https://github.com/mesosphere/marathon/pull/1066/files#diff-6876cc8f56c80c7276969fd16e24b309R400

File: src/main/resources/assets/js/components/Marathon.jsx [hunk 2]
```diff
@@ -36,11 +42,57 @@ var Marathon = React.createClass({
...
+
+    router.on("route:apps", function (appid) {
+      if (appid) {
+        if (this.state.activeAppId !== appid) {
+          this.modalDestroy();
+        }
+
+        this.setState({
+          activeAppId: appid,
+          activeApp: this.state.collection.get("/" + appid),
```
Comment (mlunoe) [inline src/main/resources/assets/js/components/Marathon.jsx]: Why do you need both here? I guess `activeAppId` should be enough. You can get the id from the collection, when you need it, or am I missing something?

File: src/main/resources/assets/js/components/Marathon.jsx [hunk 3]
```diff
@@ -385,21 +430,26 @@ var Marathon = React.createClass({
...
+        ref="modal" />
+    );
+  },
 
-    /* jshint trailing:false, quotmark:false, newcap:false */
-    /* jscs:disable disallowTrailingWhitespace, validateQuoteMarks, maximumLineLength */
-    if (this.state.modalClass === AppModalComponent) {
-      modal = (
+  routeApps: function (appid) {
+    var activeApp = this.state.activeApp;
```
Comment (mlunoe) [inline src/main/resources/assets/js/components/Marathon.jsx]: Maybe do this here?
```
var activeApp = this.state.collection.get("/" + this.state.appId);
if (activeApp) {
```

File: src/main/resources/assets/js/components/Marathon.jsx [hunk 4]
```diff
@@ -407,23 +457,34 @@ var Marathon = React.createClass({
...
+          router={this.props.router}
           ref="modal" />
       );
     }
+  },
+
+  routeNewapp: function () {
+    return (
+      <NewAppModalComponent
+        model={this.state.activeApp}
```
Comment (mlunoe) [inline src/main/resources/assets/js/components/Marathon.jsx]: Same thing here, get from collection?

File: src/main/resources/assets/js/components/Marathon.jsx [hunk 5]
```diff
@@ -163,8 +232,19 @@ var Marathon = React.createClass({
...
-  handleModalDestroy: function () {
+  modalDestroy: function () {
+    if (!this.state.modalClass) {
+      return;
+    }
+
+    var router = this.props.router;
+
+    if (router.lastRoute.hash === router.currentHash()) {
+      router.navigate("#" + this.state.activeTabId, { trigger: true });
```
Comment (mlunoe) [inline src/main/resources/assets/js/components/Marathon.jsx]: Could we remove the spaces? `{trigger: true}`

**Comment by mlunoe** [comment]
Other than that, I think it is starting to look good :)

File: src/main/resources/assets/js/models/Collection.js (hunk 1 above)
Comment (aldipower) [inline src/main/resources/assets/js/models/Collection.js]: I've tried it here:
https://github.com/mesosphere/marathon/compare/325-make-resources-linkable_cleaned...325-make-resources-linkable_updatePolling
Don't makes sense for me for this case I handle with the fetching status, because I need this line to fetch the tasks immediately:
https://github.com/mesosphere/marathon/blob/325-make-resources-linkable_cleaned/src/main/resources/assets/js/components/TaskListComponent.jsx#L30
Trigger with:
http://localhost:8080/#apps/:yourappid (reload the page to get a fresh coming in status)

If I trigger indirectly by clicking on the app in the applist I am getting multiple http-requests of the same resource, that's bad.

Don't get it how to prevent this without the fetching status..

File: src/main/resources/assets/js/components/Marathon.jsx (hunk 2 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/Marathon.jsx]: Yeah, need both, because activeAppId is used here:
https://github.com/mesosphere/marathon/pull/1066/files#diff-6876cc8f56c80c7276969fd16e24b309R180
On first page load the collection isn't ready, so this is undefined:
https://github.com/mesosphere/marathon/pull/1066#discussion-diff-23328110R77

File: src/main/resources/assets/js/components/Marathon.jsx (hunk 3 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/Marathon.jsx]: Yeah, it's a safer approach.

**Comment by aldipower** [comment]
Hehe, cool. :)

File: src/main/resources/assets/js/components/Marathon.jsx (hunk 2 above)
Comment (mlunoe) [inline src/main/resources/assets/js/components/Marathon.jsx]: That's a good point!
I'm thinking it would be cleaner to have one thing only, it seems to replicate information. If you store the app, then you can get the id `app.id`. What's your opinion?

File: src/main/resources/assets/js/components/Marathon.jsx [hunk 6]
```diff
@@ -36,11 +42,57 @@ var Marathon = React.createClass({
...
+
+    router.on("route:about", function () {
+      this.modalDestroy();
+      this.setState({
+        modalClass: AboutModalComponent
+      });
+    }.bind(this));
+
+    router.on("route:apps", function (appid) {
+      if (appid) {
```
Comment (mlunoe) [inline src/main/resources/assets/js/components/Marathon.jsx]: This could potentially be zero, let's use `appId != null`

I'm not sure if this will work, but I think it would be cleaner to move the `modalDestroy` out of the first `if`:
```
var app = this.state.collection ? this.state.collection.get("/" + appId) : null;
if (this.state.activeApp !== app) {
  this.modalDestroy();
}

if (app != null) {
  this.setState({
    activeApp: app,
    modalClass: AppModalComponent
  });
} else {
  this.activateTab("apps");
}

File: src/main/resources/assets/js/components/AppComponent.jsx [hunk 7]
```diff
@@ -3,8 +3,15 @@
...
+  propTypes: {
+    router: React.PropTypes.object.isRequired
+  },
+
   onClick: function () {
-    this.props.onClick(this.props.model);
+    this.props.router.navigate(
+      "#apps" + this.props.model.get("id"),
+      {trigger: true}
+    );
```
Comment (mlunoe) [inline src/main/resources/assets/js/components/AppComponent.jsx]: This is so nice and clean! :)

File: src/main/resources/assets/js/components/AppModalComponent.jsx [hunk 8]
```diff
@@ -31,17 +32,47 @@ var AppModalComponent = React.createClass({
...
+    this.tabs = _.reduce(tabsTemplate, function (current, tab) {
+      var id = tab.id.replace(":appid", appid);
+      if (activeTabId == null) {
+        activeTabId = id;
+      }
+      current.push({
+        id: id,
+        text: tab.text
+      });
+      return current;
```
Comment (mlunoe) [inline src/main/resources/assets/js/components/AppModalComponent.jsx]: Can we get a white space before the `return`-statement? This is a style thing we do to make it more readable.

File: src/main/resources/assets/js/components/AppModalComponent.jsx [hunk 9]
```diff
@@ -31,17 +32,47 @@ var AppModalComponent = React.createClass({
...
-    tasksFetchState: React.PropTypes.number.isRequired
+    tasksFetchState: React.PropTypes.number.isRequired,
+    router: React.PropTypes.object.isRequired
   },
 
   getInitialState: function () {
+    var appid = this.props.model.get("id");
+    var activeTabId;
+
+    this.tabs = _.reduce(tabsTemplate, function (current, tab) {
```
Comment (mlunoe) [inline src/main/resources/assets/js/components/AppModalComponent.jsx]: Let's not write to a custom variable on the object, but rather use the state

File: src/main/resources/assets/js/components/AppModalComponent.jsx [hunk 10]
```diff
@@ -31,17 +32,47 @@ var AppModalComponent = React.createClass({
...
     return {
       activeViewIndex: 0,
-      activeTabId: tabs[0].id,
+      activeTabId: activeTabId,
       selectedTasks: {}
     };
   },
 
+  componentDidMount: function () {
+    this.props.router.on("route:apps", function (appid, view) {
```
Comment (mlunoe) [inline src/main/resources/assets/js/components/AppModalComponent.jsx]: I'm not sure I understand why it is necessary to do a `router.on` function in here? Why is this not part of the state in Marathon.jsx?

Comment (mlunoe) [inline src/main/resources/assets/js/components/AppModalComponent.jsx]: I mean could this be done in Marathon.jsx's `componentDidMount` in stead and the necessary data pushed down to this component?

File: src/main/resources/assets/js/components/Marathon.jsx [hunk 11]
```diff
@@ -99,6 +151,17 @@ var Marathon = React.createClass({
         this.stopPolling();
       }
     }
+
+    var route = this.state.route;
```
Comment (mlunoe) [inline src/main/resources/assets/js/components/Marathon.jsx]: Can you explain what this is for? I'm not sure I understand where this is used?

File: src/main/resources/assets/js/components/Marathon.jsx [hunk 12]
```diff
@@ -110,10 +173,16 @@ var Marathon = React.createClass({
       error: function () {
         this.setState({fetchState: States.STATE_ERROR});
       }.bind(this),
-      reset: true,
       success: function () {
+        var state = this.state;
         this.fetchDeployments();
-        this.setState({fetchState: States.STATE_SUCCESS});
+        var activeApp = (state.activeAppId) ?
```
Comment (mlunoe) [inline src/main/resources/assets/js/components/Marathon.jsx]: Parentheses are not needed here

File: src/main/resources/assets/js/components/Marathon.jsx (hunk 6 above)
Comment (mlunoe) [inline src/main/resources/assets/js/components/Marathon.jsx]: Or maybe just use `var app = this.state.collection.get("/" + appId);` because the collection should be initialised. Or am I missing something?

File: src/main/resources/assets/js/components/Marathon.jsx (hunk 12 above)
Comment (mlunoe) [inline src/main/resources/assets/js/components/Marathon.jsx]: And same here right? Collection is initialised, so this will do: `var app = this.state.collection.get("/" + state.activeAppId);`, or am I wrong?

File: src/main/resources/assets/js/components/Marathon.jsx [hunk 13]
```diff
@@ -385,21 +430,28 @@ var Marathon = React.createClass({
...
-      modal = (
+    return (
+      <AboutModalComponent
+        onDestroy={this.modalDestroy}
+        ref="modal" />
+    );
+  },
+
+  routeApps: function () {
+    var activeApp = this.state.collection.get("/" + this.state.activeAppId);
```
Comment (mlunoe) [inline src/main/resources/assets/js/components/Marathon.jsx]: This is great!

File: src/main/resources/assets/js/components/Marathon.jsx [hunk 14]
```diff
@@ -407,22 +459,33 @@ var Marathon = React.createClass({
...
-          onDestroy={this.handleModalDestroy}
-          ref="modal" />
-      );
+    }
+  },
+
+  routeNewapp: function () {
+    return (
+      <NewAppModalComponent
+        model={this.state.collection.get("/" + this.state.activeAppId)}
```
Comment (mlunoe) [inline src/main/resources/assets/js/components/Marathon.jsx]: Why does this need a model? It creates a new one it self, as far as I can see?
https://github.com/mesosphere/marathon/blob/master/src/main/resources/assets/js/components/NewAppModalComponent.jsx#L33

File: src/main/resources/assets/js/components/NavTabsComponent.jsx [hunk 15]
```diff
@@ -42,7 +33,7 @@ var NavTabsComponent = React.createClass({
 
       return (
         <li className={tabClassSet} key={tab.id}>
-          <a href={"#" + tab.id} onClick={this.onTabClick.bind(this, tab.id)}>
+          <a href={"#" + tab.id}>
```
Comment (mlunoe) [inline src/main/resources/assets/js/components/NavTabsComponent.jsx]: Woopdi-duu! :)

File: src/main/resources/assets/js/components/Marathon.jsx (hunk 11 above)
Comment (mlunoe) [inline src/main/resources/assets/js/components/Marathon.jsx]: I'm here referring to the `router.lastRoute`

File: src/main/resources/assets/js/components/Marathon.jsx (hunk 6 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/Marathon.jsx]: There is one case, where the collection is not initialized, if you do a fresh page load with this route:
#apps/:appid
The collection isn't then ready on entry.

File: src/main/resources/assets/js/components/Marathon.jsx (hunk 2 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/Marathon.jsx]: That would be nice and cleaner, but there is one case, where the collection is initialized, but not filled with the models, if you do a fresh page load with this route:
#apps/:appid
The collection isn't then ready on entry, so I can not get the app.

File: src/main/resources/assets/js/components/AppComponent.jsx (hunk 7 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/AppComponent.jsx]: :)

File: src/main/resources/assets/js/components/AppModalComponent.jsx (hunk 9 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/AppModalComponent.jsx]: That makes sense.

File: src/main/resources/assets/js/components/AppModalComponent.jsx (hunk 10 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/AppModalComponent.jsx]: That route function is used to set the state (for the tabs) inside AppModalComponent, when the app/view changes from #apps/:appid to #apps/:appid/configure, vice versa.
I could call a callback inside Marathon.jsx that is given through the AppModal-props to set the state in AppModalComponent? But that feels like a strange wiring. :)

File: src/main/resources/assets/js/components/Marathon.jsx (hunk 12 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/Marathon.jsx]: In this case that should work! :)
But maybe there is an appId called named '/null'?

File: src/main/resources/assets/js/components/Marathon.jsx (hunk 14 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/Marathon.jsx]: Yeah, this can be removed.

File: src/main/resources/assets/js/components/NavTabsComponent.jsx (hunk 15 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/NavTabsComponent.jsx]: :)

File: src/main/resources/assets/js/components/Marathon.jsx (hunk 11 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/Marathon.jsx]: https://github.com/mesosphere/marathon/blob/325-make-resources-linkable_cleaned/src/main/resources/assets/js/components/Marathon.jsx#L242

If I close the AppModal by click on 'x', this sets the route to '#apps' or '#deployments'. But if I change the route url from '#apps/app1' to '#apps/app2' by hand in the address bar, then this should not be triggered.

File: src/main/resources/assets/js/models/Collection.js (hunk 1 above)
Comment (aldipower) [inline src/main/resources/assets/js/models/Collection.js]: Maybe I got it now. :)

File: src/main/resources/assets/js/components/AppComponent.jsx [hunk 16]
```diff
@@ -3,8 +3,15 @@
...
 
 var AppComponent = React.createClass({
+  propTypes: {
+    router: React.PropTypes.object.isRequired
+  },
+
   onClick: function () {
-    this.props.onClick(this.props.model);
+    this.props.router.navigate(
+      "#apps" + this.props.model.get("id"),
```
Comment (pyronicide) [inline src/main/resources/assets/js/components/AppComponent.jsx]: This is dependent on using hashes, it would be nice if we left that up to the router to handle whether it is fragments or push state.

Comment (pyronicide) [inline src/main/resources/assets/js/components/AppComponent.jsx]: Shouldn't the routes be following as closely as possible to a REST style interface? I'd expect to see something more like:

`apps/:id`

File: src/main/resources/assets/js/components/AppListComponent.jsx [hunk 17]
```diff
@@ -12,17 +12,13 @@ var AppListComponent = React.createClass({
 
   propTypes: {
     collection: React.PropTypes.object.isRequired,
-    onSelectApp: React.PropTypes.func.isRequired
+    router: React.PropTypes.object.isRequired
```
Comment (pyronicide) [inline src/main/resources/assets/js/components/AppListComponent.jsx]: Passing a router all the way down the tree is the same thing as creating a global. It suggests that we've not encapsulated things correctly and need to come at it from a different perspective.

File: src/main/resources/assets/js/components/AppModalComponent.jsx [hunk 18]
```diff
@@ -31,17 +33,49 @@ var AppModalComponent = React.createClass({
...
-    tasksFetchState: React.PropTypes.number.isRequired
+    tasksFetchState: React.PropTypes.number.isRequired,
+    router: React.PropTypes.object.isRequired
   },
 
   getInitialState: function () {
+    var appid = this.props.model.get("id");
+    var activeTabId;
+
+    var tabs = _.reduce(tabsTemplate, function (current, tab) {
```
Comment (pyronicide) [inline src/main/resources/assets/js/components/AppModalComponent.jsx]: Again, shouldn't this be getting handled at a higher level? The component shouldn't be doing raw munging of the whole URL. It should just have a switch as to whether something is being shown or not.

File: src/main/resources/assets/js/components/AppModalComponent.jsx [hunk 19]
```diff
@@ -31,17 +33,49 @@ var AppModalComponent = React.createClass({
...
+      selectedTasks: {},
+      tabs: tabs
     };
   },
 
+  componentDidMount: function () {
+    this.props.handleSetAppView (function (appid, view) {
+      if (appid && this.isMounted()) {
+        this.setState({
+          activeTabId: "apps/" + appid + (view ? "/" + view : "")
```
Comment (pyronicide) [inline src/main/resources/assets/js/components/AppModalComponent.jsx]: Needs to be, at least, a common lib instead of being spread across the function.

File: src/main/resources/assets/js/components/AppModalComponent.jsx [hunk 20]
```diff
@@ -31,17 +33,49 @@ var AppModalComponent = React.createClass({
...
-      selectedTasks: {}
+      activeTabId: activeTabId,
+      selectedTasks: {},
+      tabs: tabs
     };
   },
 
+  componentDidMount: function () {
+    this.props.handleSetAppView (function (appid, view) {
+      if (appid && this.isMounted()) {
```
Comment (pyronicide) [inline src/main/resources/assets/js/components/AppModalComponent.jsx]: Why is `this.isMounted()` required?

File: src/main/resources/assets/js/components/AppModalComponent.jsx [hunk 21]
```diff
@@ -31,17 +33,49 @@ var AppModalComponent = React.createClass({
...
+        this.setState({
+          activeTabId: "apps/" + appid + (view ? "/" + view : "")
+        });
+      }
+    }.bind(this));
+
+    this.setState({
```
Comment (pyronicide) [inline src/main/resources/assets/js/components/AppModalComponent.jsx]: I've had quite a few nasty surprises doing this (you'll get some ugly double renders and can occasionally get into a loop).

File: src/main/resources/assets/js/components/AppModalComponent.jsx [hunk 22]
```diff
@@ -167,8 +195,8 @@ var AppModalComponent = React.createClass({
         <TogglableTabsComponent className="modal-body modal-body-no-top"
             activeTabId={this.state.activeTabId}
             onTabClick={this.onTabClick}
-            tabs={tabs} >
-          <TabPaneComponent id="tasks">
+            tabs={this.state.tabs} >
+          <TabPaneComponent id={"apps" + model.get("id")}>
```
Comment (pyronicide) [inline src/main/resources/assets/js/components/AppModalComponent.jsx]: Common method.

File: src/main/resources/assets/js/components/AppModalComponent.jsx [hunk 23]
```diff
@@ -189,7 +217,7 @@ var AppModalComponent = React.createClass({
             </StackedViewComponent>
           </TabPaneComponent>
           <TabPaneComponent
-            id="configuration"
+            id={"apps" + model.get("id") + "/configuration"}
```
Comment (pyronicide) [inline src/main/resources/assets/js/components/AppModalComponent.jsx]: Same as above.

File: src/main/resources/assets/js/components/AppVersionComponent.jsx [hunk 24]
```diff
@@ -27,7 +27,7 @@ var AppVersionComponent = React.createClass({
   },
 
   handleSubmit: function (event) {
-    if (_.isfunction (this.props.onRollback)) {
+    if (_.isFunction (this.props.onRollback)) {
```
Comment (pyronicide) [inline src/main/resources/assets/js/components/AppVersionComponent.jsx]: Did this just not work before?

File: src/main/resources/assets/js/components/Marathon.jsx [hunk 25]
```diff
@@ -2,6 +2,7 @@
 
 var Mousetrap = require("mousetrap");
 var React = require("react/addons");
+var _ = require("underscore");
```
Comment (pyronicide) [inline src/main/resources/assets/js/components/Marathon.jsx]: Let's keep the requires alphabetized.

File: src/main/resources/assets/js/components/Marathon.jsx [hunk 26]
```diff
@@ -36,11 +43,61 @@ var Marathon = React.createClass({
       deploymentsFetchState: States.STATE_LOADING,
       fetchState: States.STATE_LOADING,
       modalClass: null,
+      route: null,
+      setAppView: function () {},
```
Comment (pyronicide) [inline src/main/resources/assets/js/components/Marathon.jsx]: Buh?

File: src/main/resources/assets/js/components/Marathon.jsx [hunk 27]
```diff
@@ -36,11 +43,61 @@ var Marathon = React.createClass({
...
       tasksFetchState: States.STATE_LOADING
     };
   },
 
   componentDidMount: function () {
+    var router = this.props.router;
+
+    router.on("route", function (route, params) {
```
Comment (pyronicide) [inline src/main/resources/assets/js/components/Marathon.jsx]: Each of these should be their own methods.

File: src/main/resources/assets/js/components/Marathon.jsx [hunk 28]
```diff
@@ -36,11 +43,61 @@ var Marathon = React.createClass({
...
+          activeApp: this.state.collection.get("/" + appid),
+          modalClass: AppModalComponent,
+          activeAppView: view
+        });
+      } else {
+        this.activateTab("apps");
+      }
+    }.bind(this));
+
+    router.on("route:deployments", function () {
```
Comment (pyronicide) [inline src/main/resources/assets/js/components/Marathon.jsx]: router.on("route:deployments", _.bind(this.activateTab, None, "deployments"));

File: src/main/resources/assets/js/components/Marathon.jsx [hunk 29]
```diff
@@ -51,23 +108,23 @@ var Marathon = React.createClass({
 
     Mousetrap.bind("esc", function () {
       if (this.refs.modal != null) {
-        this.refs.modal.destroy();
+        this.modalDestroy();
       }
     }.bind(this));
 
     Mousetrap.bind("c", function () {
```
Comment (pyronicide) [inline src/main/resources/assets/js/components/Marathon.jsx]: Can we just remove this binding? It gets in the way of copy/paste and is a usability issue. Happy to handle it in another PR.

File: src/main/resources/assets/js/components/Marathon.jsx [hunk 30]
```diff
@@ -78,26 +135,31 @@ var Marathon = React.createClass({
...
+    }
+
+    if (this.state.activeAppId) {
+      this.state.setAppView(this.state.activeAppId, this.state.activeAppView);
+    }
+
+    var route = this.state.route;
+    var router = this.props.router;
+
+    if (route) {
```
Comment (pyronicide) [inline src/main/resources/assets/js/components/Marathon.jsx]: Why is this here at all? We shouldn't care about the last state.

File: src/main/resources/assets/js/components/Marathon.jsx [hunk 31]
```diff
@@ -110,10 +172,16 @@ var Marathon = React.createClass({
       error: function () {
         this.setState({fetchState: States.STATE_ERROR});
       }.bind(this),
-      reset: true,
       success: function () {
+        var state = this.state;
```
Comment (pyronicide) [inline src/main/resources/assets/js/components/Marathon.jsx]: Why? Doesn't look like this is needed?

File: src/main/resources/assets/js/components/Marathon.jsx [hunk 32]
```diff
@@ -110,10 +172,16 @@ var Marathon = React.createClass({
...
         this.fetchDeployments();
-        this.setState({fetchState: States.STATE_SUCCESS});
+        var activeApp = state.activeAppId ?
```
Comment (pyronicide) [inline src/main/resources/assets/js/components/Marathon.jsx]: Seems like the right thing for a helper function.

File: src/main/resources/assets/js/components/Marathon.jsx [hunk 33]
```diff
@@ -159,12 +227,29 @@ var Marathon = React.createClass({
     }
   },
 
+  handleSetAppView: function (setAppView) {
```
Comment (pyronicide) [inline src/main/resources/assets/js/components/Marathon.jsx]: Why is this here along with passing the router down? Shouldn't it all come from the top?

File: src/main/resources/assets/js/components/Marathon.jsx [hunk 34]
```diff
@@ -338,68 +423,51 @@ var Marathon = React.createClass({
...
+        onDestroy={this.modalDestroy}
+        ref="modal" />
+    );
+    /* jshint trailing:true, quotmark:true, newcap:true */
+    /* jscs:enable disallowTrailingWhitespace, validateQuoteMarks, maximumLineLength */
+  },
+
+  routeApps: function () {
+    var activeApp = this.state.collection.get("/" + this.state.activeAppId);
+    if (activeApp) {
```
Comment (pyronicide) [inline src/main/resources/assets/js/components/Marathon.jsx]: if (!activeApp) { return }

    return (<AppModalComponent>)

File: src/main/resources/assets/js/components/Marathon.jsx [hunk 35]
```diff
@@ -338,68 +423,51 @@ var Marathon = React.createClass({
...
+      return (
         <AppModalComponent
           activeTask={this.state.activeTask}
           appVersionsFetchState={this.state.appVersionsFetchState}
           destroyApp={this.destroyApp}
           fetchTasks={this.fetchTasks}
           fetchAppVersions={this.fetchAppVersions}
-          model={this.state.activeApp}
-          onDestroy={this.handleModalDestroy}
+          handleSetAppView={this.handleSetAppView}
```
Comment (pyronicide) [inline src/main/resources/assets/js/components/Marathon.jsx]: The sheer number of params here suggests we're doing it wrong (tm). The encapsulation should be way better here.

File: src/main/resources/assets/js/components/AppComponent.jsx (hunk 16 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/AppComponent.jsx]: Yes, 2 good points.
Now using encodeURIComponent to allow route 'apps//sleep30', it becomes 'apps/%2Fsleep30'.
'/sleep30' is an appid.

File: src/main/resources/assets/js/components/AppListComponent.jsx (hunk 17 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/AppListComponent.jsx]: Yes, I also have a bigger refactoring task in mind. We should talk about that in the future all together.

File: src/main/resources/assets/js/components/AppModalComponent.jsx (hunk 18 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/AppModalComponent.jsx]: Yes, but need unique ids at the moment for the NavTabComponent.
Should be a seperate refactoring task.

File: src/main/resources/assets/js/components/AppModalComponent.jsx (hunk 19 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/AppModalComponent.jsx]: What do you mean with that?

File: src/main/resources/assets/js/components/AppVersionComponent.jsx (hunk 24 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/AppVersionComponent.jsx]: Shouldn't, Underscores function is named isFunction.

File: src/main/resources/assets/js/components/AppModalComponent.jsx (hunk 20 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/AppModalComponent.jsx]: That prevents the double renders you've mentioned below.
But I've found a cleaner solution with componentWillUnmount.

File: src/main/resources/assets/js/components/AppModalComponent.jsx (hunk 22 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/AppModalComponent.jsx]: What do you mean with that?

File: src/main/resources/assets/js/components/Marathon.jsx (hunk 29 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/Marathon.jsx]: Yeah, maybe we can discuss other bindings, too.

File: src/main/resources/assets/js/components/Marathon.jsx (hunk 31 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/Marathon.jsx]: this.state is used twice, so I can save one object lookup.
In a larger scale this gains performance and readability.

File: src/main/resources/assets/js/components/Marathon.jsx (hunk 28 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/Marathon.jsx]: Cool. :)

File: src/main/resources/assets/js/components/Marathon.jsx (hunk 27 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/Marathon.jsx]: This is cleaner, no bind needed.

File: src/main/resources/assets/js/components/Marathon.jsx (hunk 33 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/Marathon.jsx]: Yes, in an ideal case. Maybe we should handle this issue of passing things down in a seperate pull request.
See my comment here:
https://github.com/mesosphere/marathon/pull/1066#discussion-diff-23438066
This is a result of getting the route function out of the AppModalComponent.

File: src/main/resources/assets/js/components/Marathon.jsx (hunk 34 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/Marathon.jsx]: Yeah, makes sense again.
https://github.com/mesosphere/marathon/pull/990#discussion-diff-23279372

File: src/main/resources/assets/js/components/Marathon.jsx (hunk 30 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/Marathon.jsx]: This is a little bit tricky. We need this to populate the tasks for the AppModal correctly. Because the AppModal can be opened on various states.

File: src/main/resources/assets/js/components/Marathon.jsx (hunk 32 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/Marathon.jsx]: Sorry, what do you mean?

File: src/main/resources/assets/js/components/Marathon.jsx (hunk 35 above)
Comment (aldipower) [inline src/main/resources/assets/js/components/Marathon.jsx]: Yes, I am with you.
We should think about a better encapsulation, doing structural refactoring, making Marathon.jsx more leightweight, bringing function to the place where it is needed, clean up the render cycle, etc.
Let's do this in a seperate issue and discuss about it all togther.

File: src/main/resources/assets/js/components/AppModalComponent.jsx (hunk 18 above)
Comment (pyronicide) [inline src/main/resources/assets/js/components/AppModalComponent.jsx]: I'm okay with it being a separate task as long as this doesn't get merged until that happens. Having routing at this level has way too many repercussions and will be difficult to remove in the future.

File: src/main/resources/assets/js/components/Marathon.jsx (hunk 30 above)
Comment (pyronicide) [inline src/main/resources/assets/js/components/Marathon.jsx]: The fact that this is possible means that we've got the high level architecture right. Let's get that fixed instead of continuing to monkey-patch it.

File: src/main/resources/assets/js/components/Marathon.jsx (hunk 31 above)
Comment (pyronicide) [inline src/main/resources/assets/js/components/Marathon.jsx]: You're getting neither here. This is assigning a reference and now I need to think about whether you've got a local state or are still using `this.state`.

File: src/main/resources/assets/js/components/Marathon.jsx (hunk 32 above)
Comment (pyronicide) [inline src/main/resources/assets/js/components/Marathon.jsx]: `this.getActiveApp()`

File: src/main/resources/assets/js/components/Marathon.jsx (hunk 33 above)
Comment (pyronicide) [inline src/main/resources/assets/js/components/Marathon.jsx]: Well, our code should be the ideal case.
